streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```
@@ -570,13 +591,34 @@ public StreamsConfig(final Map<?, ?> props) {
      * @return Map of the producer configuration.
      */
     public Map<String, Object> getProducerConfigs(final String clientId) {
+        final Map<String, Object> clientProvidedProps = getClientPropsWithPrefix(PRODUCER_PREFIX, ProducerConfig.configNames());
+
         // generate producer configs from original properties and overridden maps
         final Map<String, Object> props = new HashMap<>(PRODUCER_DEFAULT_OVERRIDES);
         props.putAll(getClientPropsWithPrefix(PRODUCER_PREFIX, ProducerConfig.configNames()));
 
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, originals().get(BOOTSTRAP_SERVERS_CONFIG));
         // add client id with stream client id prefix
         props.put(CommonClientConfigs.CLIENT_ID_CONFIG, clientId + "-producer");
+        if (eosEnabled) {
+            if (clientProvidedProps.containsKey(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG)) {
+                throw new ConfigException("Unexpected user-specified consumer config " + ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG
+                    + "; because " + PROCESSING_GUARANTEE_CONFIG + " is set to '" + EXACTLY_ONCE + "' producer will always have idempotency enabled.");
+            }
+            props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+
+            if (clientProvidedProps.containsKey(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION)) {
+                throw new ConfigException("Unexpected user-specified consumer config " + ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION
```

This is actually a big deal. Should have been in streams KIP. Will impact performance a lot.

---

Are there some numbers on this? It would be useful to know.

---

This setting is part of KIP-98 -- we just follow what the producer client dictates if we want to use idempotency (and transactions).

It think there are numbers (\cc @apurvam can you point to some benchmark results?)

---

The numbers are here: https://docs.google.com/spreadsheets/d/1dHY6M7qCiX-NFvsgvaE0YoVdNq26uA8608XIh_DUpI4/edit#gid=1276994626

As you can see ,setting inflights to 1 from the default of 5 does have a perf impact of 20% on those tests. But also note that compression was enabled and batch sizes were small.  

I think that if you tune your batches to be larger, the impact will be ameliorated. For instance, if you commit transactions every 500ms, your linger ms could be 500ms as well for maximum batches. With larger batches, you don't get much benefit from multiple inflights.

Finally, we can't guarantee in order delivery without this setting even before KIP-98. Are all streams operators currently symmetric? Otherwise they would already be buggy.

---

this should be set only when we are in eos mode right?

---

Yes.

---

That is what we do: See line 603: `if (eosEnabled) {`

---

Ok thanks.

---

@mjsax could we run the streams benchmarks to see impact? Thanks.

---

To clarify, I am aware of the benchmarks that we did for KIP-98. I was asking if Eno had quantified the performance impact of this config in the context of Streams benchmarks (Eno did some work on that recently).

---

I haven't quantified that. Now is a good time to do that :)

---

Do we need to override `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION` here, since it is already taken care of inside `KafkaProducer#configureInflightRequests` that, if idemponency is set, max.in.flight.request must be 1 or otherwise exception thrown?

---

Even if producer does check this config, it would be better if Streams would do this check too, and if we go with "overwrite and log" approach, sets the value to `1`.

Btw: if we go with "throw exception" approach, I would keep this check in Streams, too. We should move all checks into the constructor to fail faster. That will simplify debugging for users.